content/post/2021-10-07-input-checking/index.Rmd

```diff
@@ -0,0 +1,301 @@
+Let's imagine we have written a short function to compute the geometric mean:
+
+```{r}
+geometric_mean <- function(...) {
+  
+  return(prod(...)^(1/...length()))
+  
+}
+```
+
+When you tested the function yourself, everything seemed fine:
+
+```{r}
+geometric_mean(2, 8)
+
+geometric_mean(4, 1, 1/32)
+```
+
+But a different person using your function might expose it to situations it was not prepared to handle, resulting in cryptic errors or undefined behaviour:
+
+```{r, error = TRUE}
+# Input with factors instead of numerics
+geometric_mean(factor(2), 8)
+
+# Input with negative values
+geometric_mean(-1, 5)
+
+# Input with NAs
+geometric_mean(2, 8, NA)
+```
+
+Or worse, it could give an incorrect output:
+
+```{r}
+geometric_mean(c(2, 8))
+```
+
+Because of this, you need to make sure you return clear errors whenever your functions receives input it was not designed for.
+In this blog post, we review a range of approaches to help you check your function inputs and discuss some potential future developments.
+
+## Pre-requisite: thoroughly document your argument types
+
+You can notice from the simple example above that it's easy to pass invalid inputs to the `geometric_mean()` function because we didn't provide any documentation on what is or isn't a valid input.
+We won't go into details here but the [roxygen2](https://roxygen2.r-lib.org/) package provides a convenient way to generate documentation for R functions.
+Try to be as precise as possible when describing the required format for your inputs [^1].
+
+[^1]: [Some package developers even developed their own standardized way to document argument types and length](https://github.com/r-lib/withr/commit/42e503092046705f30032cb3a321d64b0e9383d4).
+    But there is currently no standard shared across the R community.
+
+```{r}
+#' @param name A character of length one with the name of the person to greet
+say_hello <- function(name) {
+  stopifnot(is.character(name))
+  paste("Hello", name)
+}
+```
+
+Adding any kind of argument checking in the absence of good documentation would be vain and very frustrating for your users as they would have to figure out what is or isn't valid by trial and error.
+
+## Checking function inputs using base R [^2]
+
+[^2]: Note that these base functions have equivalent in the tidyverse with a more consistent design and coloured output.
+    `match.arg()`'s equivalent is `rlang::arg_match()` and `stopifnot()`'s
+
+### `match.arg()`
+
+If the input can only take specific values, the base function `match.arg()` can also prove useful:
+
+```{r, error = TRUE}
+match.arg(arg = "R", choices = c("R", "python"))
+
+match.arg(arg = "javascript", choices = c("R", "python"))
+```
+
+But the real power of the `match.arg()` function comes from the fact that `choices` can be automatically obtained in the context of a function:
+
+```{r, error = TRUE}
+choose_language <- function(language = c("R", "python")) {
```

Comment:
One thing that's not obvious to me when using this pattern is what the default value is. Could you add a sentence about this?

Reply:
(I know it _should_ be obvious to me :sweat_smile: :see_no_evil: )